### PR TITLE
refactor: stake withdraw logic

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
The stake withdraw logic is hard to follow

#### Summary of Changes
No functional changes or error variants should change as a result of this change.

- https://github.com/anza-xyz/agave/commit/eccb68d1c2b7373037fbac2ae4286e6109949bc5 straightforward logical refactor
- https://github.com/anza-xyz/agave/commit/cebea2f892bb0985d76966c590684fdef4912f87 pull out full withdrawal special case and check it first. Note that if `lamports == stake_account.get_lamports()`, then it's impossible for `checked_add(lamports, reserve)?` to fail
- https://github.com/anza-xyz/agave/commit/84a5994c3f638eca6f224f48ea6f20b32222889c straightforward logical refactor

Fixes #

### CODE REVIEW

**Review by 2501babe — APPROVED** [review]

**Comment by 2501babe** [comment]
i also disliked that code but merely made it more descriptive in bpf
```rust
        let withdraw_lamports_and_reserve = checked_add(withdraw_lamports, reserve)?;
        let stake_account_lamports = source_stake_account_info.lamports();

        // if the stake is active, we mustn't allow the account to go away
        if is_staked && withdraw_lamports_and_reserve > stake_account_lamports {
            return Err(ProgramError::InsufficientFunds);
        }       

        // a partial withdrawal must not deplete the reserve
        if withdraw_lamports != stake_account_lamports
            && withdraw_lamports_and_reserve > stake_account_lamports
        {       
            assert!(!is_staked);
            return Err(ProgramError::InsufficientFunds);
        }       

        // Deinitialize state upon zero balance
        if withdraw_lamports == stake_account_lamports {
            set_stake_state(source_stake_account_info, &StakeStateV2::Uninitialized)?;
        }
```

**Comment by jstarry** [comment]
@lheeger-jump can someone from your team review this change please? It's not that important to merge if you feel strongly against it but this refactoring makes the behavior and error cases for stake withdraws a lot clearer.

**Comment by jstarry** [comment]
> Note that if `lamports == stake_account.get_lamports()`, then it's impossible for `checked_add(lamports, reserve)?` to fail

And to elaborate on the above note from the PR description. The reason this is impossible is because it's not possible in the SVM to even temporarily give an account near `u64::MAX` lamports because the runtime does lamport accounting checks before and after every instruction boundary which prohibit creating lamports out of thin air.

**Comment by jstarry** [comment]
fyi @topointon-jump
